### ISSUE TITLE
[backend]  Unable to create directory in Minio when using Artifacts: Permission denied

### ISSUE THREAD

**Issue by jmaunon** [kubeflow/pipelines#10397]
Hi Developers

I have tried to create a simple pipeline using and transfering data using "built-in" `artifacts` approach without success.
Difficult to say what is hapenning but I have found similar issues in other threads. 

Please, if you know a manual patch, let us know. I see `artifacts` a core solution/approach.

cc: @juliusvonkohout , @chensun 

I am aware that there are some issues related, but I do not see a final solution or alternative patch. See: #6530 , https://github.com/kubeflow/manifests/issues/2573, https://github.com/kubeflow/pipelines/issues/7629

### Environment

*  How did you deploy Kubeflow Pipelines (KFP): [https://github.com/kubeflow/manifests](https://github.com/kubeflow/manifests) latest tag: v.1.8.0

*  KFP version: 
According to the `Readme` from manifest repo: KFP 2.0.3
*  KFP SDK version:  2.6.0

### Steps to reproduce

I get a `permission denied` error when using `Artifacts`. 

Snippet of code:

```
@dsl.component(base_image="kubeflownotebookswg/jupyter-pytorch-full:v1.8.0-rc.0")
def download_data(test_path: Output[Dataset]):
    
    import torch
    
    from torchvision.transforms import ToTensor
    from torchvision.datasets import MNIST
    
    mnist_test  = MNIST(".", download=True, train=False, transform=ToTensor())
        
    with open(test_path.path, "wb") as f:
        torch.save(mnist_test,f)

@dsl.pipeline(
    name='mnist',
    description='Detect digits',
)
def run():
    step_1 = download_data()

client.create_run_from_pipeline_func(run)
```

Associated logs:

`failed to execute component: unable to create directory "/minio/mlpipeline/v2/artifacts/mnist/43f760f9-b638-4129-87fe-602e24076beb/download-data" for output artifact "test_path": mkdir /minio: permission denied`

### Expected result

Work without issus

### Materials and Reference

---

Impacted by this bug? Give it a 👍.

**Comment by juliusvonkohout** [comment]
Please use the final 1.8 image, not jupyter-pytorch-full:v1.8.0-rc.0 and join the biweekly KFP meeting to discuss this.

**Comment by juliusvonkohout** [comment]
You should also try to update from KFP 2.0.3 to 2.0.5 first.

**Comment by jmaunon** [comment]
Thans for the reply @juliusvonkohout . I write here my findings:
* I have already tested with `jupyter-pytorch-full:v1.8.0-rc.0` and `jupyter-pytorch-full:v1.8.0` and the problem persist.
* I have not updated to `2.0.5`, so I cannot confirm if the error is fixed or not (I do not think so).

For any readers,  I did not understand the explanation of #6530  but:
* If I use `base_image=python:3.10` the pipelines executes without problem because the user of the `docker` seems to be `root`. See associated  [dockerfiles](https://github.com/docker-library/python/tree/2d31ccc9f8487908ded7944a54b8e923eff9ad1f/3.10)
* If use `base_image=kubeflownotebookswg/jupyter-pytorch-full:v1.8.0` the pipeline raises the `permission denied problem` and i I see in the associated dockerfile that the user of the `docker image` is not root. See asociated [dockerfile](https://github.com/kubeflow/kubeflow/blob/master/components/example-notebook-servers/jupyter/Dockerfile#L42)

**Comment by juliusvonkohout** [comment]
@rimolive this might be something to track

**Comment by rimolive** [comment]
We have an open PR for that #10538.

**Comment by majuss** [comment]
/reopen
This issue still persists

**Comment by thesuperzapper** [comment]
This issue is actually because Kubeflow Pipelines requires that component containers run as root, the container you have chosen `kubeflownotebookswg/jupyter-pytorch-full:v1.8.0-rc.0` runs as non-root.

There is a PR to fix this issue by mounting `emptyDir` volumes at the `/minio` and other paths, but that will need to be reviewed:
- https://github.com/kubeflow/pipelines/pull/10538

@chensun @HumairAK @Tomcli we definitely need to prioritize fixing this issue, because it's pretty bad to have a hard requirement on root container images.

**Comment by thesuperzapper** [comment]
I also want to say that the lack of `securityContext` support is related to this, because if we had it, it would provide a possible workaround:
- https://github.com/kubeflow/pipelines/issues/9783 

That is, if users could set the [Pod `securityContext`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#securitycontext-v1-core), they could set `runAsUser: 0` to override the UID of images which don't run as root by default.

**Comment by droctothorpe** [comment]
We're running into this now. All our end user containers run as non-root to optimize security. This is a pretty universal expectation at any security sensitive company.

**Comment by droctothorpe** [comment]
For anyone else running into this, we found a short-term workaround using kyverno that's not contingent on this PR being merged. Huge shout out to @moorthy156 for implementing it lightning fast. Just update the `mountPath` to `minio` or `gcs` or whatever else you need it to be.

```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: add-volume-mount-pipelineroot
spec:
  background: true
  failurePolicy: Ignore
  rules:
  - match:
      any:
      - resources:
          kinds:
          - Pod
          namespaceSelector:
            matchLabels:
              app.kubernetes.io/part-of: "kubeflow-profile"
          selector:
            matchExpressions:
            - key: pipelines.kubeflow.org/v2_component
              operator: In
              values:
              - "true"
    mutate:
      patchStrategicMerge:
        spec:
          volumes:
          - name: pipelineroot
          containers:
          - (name): main | wait
            volumeMounts:
            - mountPath: /s3
              name: pipelineroot
            env:
            - name: AWS_REGION
              value: us-east-1
    name: add-volume-mount-pipelineroot
    preconditions:
      all:
      - key: '{{ request.operation }}'
        operator: Equals
        value: CREATE
```

**Comment by thesuperzapper** [comment]
Just wanted to update everyone that there is a new PR being worked on that will fix this issue:

- https://github.com/kubeflow/pipelines/pull/10857

**Comment by thesuperzapper** [comment]
@chensun @james-jwu @zijianjoy can we please cherry-pick https://github.com/kubeflow/pipelines/pull/10857 into the 2.2 branch, and cut a 2.2.1 release with this fix?

This is a very important issue, as it prevents non-root containers from working in pipeline steps, which stops many people adopting Kubeflow Pipelines.

**Comment by juliusvonkohout** [comment]
> @chensun @james-jwu @zijianjoy can we please cherry-pick #10857 into the 2.2 branch, and cut a 2.2.1 release with this fix?
> 
> This is a very important issue, as it prevents non-root containers from working in pipeline steps, which stops many people adopting Kubeflow Pipelines.

We can also do a follow up Kubeflow 1.9.1, but one way or the other we need a new release of KFP.

**Comment by thesuperzapper** [comment]
This should have been resolved by https://github.com/kubeflow/pipelines/pull/10857 in 2.3.0

/close
